src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs

```diff
@@ -441,15 +449,31 @@ private static string GetDisplayGroup(FieldInfo field)
     }
 
     internal static bool IsRequired(ValidationMetadataProviderContext context)
+    {
+        var nullability = CreateNullabilityInfo(context);
+        var isOptional = nullability != null && nullability.ReadState != NullabilityState.NotNull;
+        return !isOptional;
+    }
+
+    internal static bool HasRequiredGeneric(ValidationMetadataProviderContext context)
+    {
+        var nullability = CreateNullabilityInfo(context);
+        if (nullability is { GenericTypeArguments: { Length: 1 } })
```

Comment:
Does this work for arbitrarily nested thingies e.g. `List<KeyValuePair<string, Person>>`? Also arrays? It might be cool to say it only works for certain shapes of types and collections, but we should discuss that with the team so like everyone's on board or in case there are better ways to tackle this.